### DISCUSSION TITLE
Is there a way to list all the validators that the validator client is signing for?

### DISCUSSION THREAD

**Discussion by rishabh-vasudevan19** [Consensys/teku#7543] [Q&A]
I had a usecase in which I need to list all the validators, it could be an api endpoint, or a metrics page where the validator index or address is mentioned.

**Comment by benjaminion** [comment]
The `/teku/v1/beacon/proposers_data` endpoint looks as if it should do what you want. Docs are [here](https://consensys.github.io/teku/#tag/Experimental/operation/getProposersData). I think the returned `registered_validators` array should be the list you need, but I'm not in a position to test it right now.

**Reply by rishabh-vasudevan19** [reply]
Thank you, I'll test it out myself

**Comment by benjaminion** [comment]
More info from @tbenr :

> `registered_validators` are the list of validators that has been registered to use the builder flow. `prepared_proposers` are the list of validators that the beacon node expect to be proposers using the beacon node. So I think he should look to `prepared_proposers` .
Normally they are the same set if all validators are configured to use MEV-Boost.
>
> Also note that if the same BN is used by multiple VCs, the union of all validators will be exposed here. So depends by the setup if this approach fits the needs.

**Reply by rishabh-vasudevan19** [reply]
Got it, Thanks

**Comment by tbenr** [comment]
@rishabh-risk-harbor there is also another way which is looking at metrics. If you enable metrics on validator client you can then access `validator_local_validator_counts` metric
![image](https://github.com/Consensys/teku/assets/15999009/df51db4a-373e-41b2-9cd2-da4ca123f265)
all statuses starting with "active_" will give you validators number currently performing duties. If none are exiting or slashed, all validators are under `active_ongoing` status.

**Reply by rishabh-vasudevan19** [reply]
Do you know of a way to get the indexes or addresses of active validators in the validator client?

**Comment by rishabh-vasudevan19** [comment]
Hey Guys,
I have a use case in which I need the validators that the validator client is signing for. Would it be possible that I make a PR which exposes this kind of data. Something similar to https://lighthouse-book.sigmaprime.io/api-vc-endpoints.html#:~:text=the%20authorization%20token.-,GET%20/lighthouse/validators,-List%20all%20validators

**Comment by benjaminion** [comment]
Does the `/teku/v1/beacon/proposers_data` endpoint not give you the information you need? What would need to be added to that to meet your use case?

PRs are welcome if they address a gap, but we'd not want to duplicate the existing functionality unnecessarily.

**Reply by rishabh-vasudevan19** [reply]
@benjaminion As you had mentioned in your second reply "Also note that if the same BN is used by multiple VCs, the union of all validators will be exposed here. So depends by the setup if this approach fits the needs". In my use case I need just the ones that are being signed for by the validator client and not the union of validators that are using the same BN.

**Reply by rishabh-vasudevan19** [reply]
@benjaminion any views on this?

**Reply by tbenr** [reply]
Another option is to enable validator client API https://docs.teku.consensys.net/reference/rest#enable-the-validator-client-api and consume the key manager api to list keys handled by the VC. https://ethereum.github.io/keymanager-APIs/

**Reply by Nar035** [reply]
Thanks, was 
searching for this.
